### PR TITLE
BTAS development branch

### PR DESCRIPTION
**Pull request by kmp5VT**
Some optimizations and other small changes to BTAS. Fix CP_DF_ALS, mixed contractions for component tensors with more than 3 modes was implemented improperly, improve efficiency of code by storing partial grammians, remove the explicit flatten call in tucker.h

### CODE REVIEW

File: btas/generic/converge_class.h [hunk 1]
```diff
@@ -319,6 +319,10 @@ namespace btas {
       return final_fit_;
     }
 
+    double get_norm(std::vector<Tensor> & btas_array){
```
Comment (evaleev) [inline btas/generic/converge_class.h:322]: `const&`

File: btas/generic/cp.h [hunk 2]
```diff
@@ -116,17 +133,15 @@ namespace btas{
...
     using ord_t = typename range_traits<typename Tensor::range_type>::ordinal_type;
 
     /// Create a generic CP object that stores the factor matrices,
     /// the number of iterations and the number of dimensions of the original
     /// tensor
     /// \param[in] ndim number of modes in the reference tensor.
-    CP(size_t dims) : num_ALS(0) {
-      ndim = dims;
-    }
+    CP(size_t dims) : num_ALS(0) { ndim = dims; }
```
Comment (evaleev) [inline btas/generic/cp.h:144]: prefer to initialize in the initializer list

File: btas/generic/cp.h [hunk 3]
```diff
@@ -158,15 +173,15 @@ namespace btas{
     /// error between exact and approximate tensor, -1 if calculate_epsilon =
     /// false && ConvClass != FitCheck.
 
-    double compute_rank(ind_t rank, ConvClass &converge_test, ind_t step = 1,
-                        bool SVD_initial_guess = false, ind_t SVD_rank = 0, ind_t max_als = 1e4,
-                        bool fast_pI = false, bool calculate_epsilon = false, bool direct = true) {
+    double compute_rank(ind_t rank, ConvClass &converge_test, ind_t step = 1, bool SVD_initial_guess = false,
```
Comment (evaleev) [inline btas/generic/cp.h:176]: Update the dox, they don't seem to be in sync (e.g. default for `fast_pI` is now `false`, not `true`, etc.

File: btas/generic/cp.h [hunk 4]
```diff
@@ -191,13 +206,12 @@ namespace btas{
     /// \returns 2-norm
     /// error between exact and approximate tensor, -1 if calculate_epsilon =
     /// false && ConvClass != FitCheck.
-    double compute_rank_random(ind_t rank, ConvClass &converge_test, ind_t max_als = 1e4,
-                               bool fast_pI = false, bool calculate_epsilon = false, bool direct = true) {
+    double compute_rank_random(ind_t rank, ConvClass &converge_test, ind_t max_als = 1e4, bool fast_pI = false,
```
Comment (evaleev) [inline btas/generic/cp.h:209]: dox

File: btas/generic/cp.h [hunk 5]
```diff
@@ -231,9 +245,9 @@ namespace btas{
     /// \returns 2-norm
     /// error between exact and approximate tensor, -1 if calculate_epsilon =
     /// false && ConvClass != FitCheck.
-    double compute_error(ConvClass &converge_test, double tcutCP = 1e-2, ind_t step = 1,
-                         ind_t max_rank = 1e5, bool SVD_initial_guess = false, ind_t SVD_rank = 0,
-                         ind_t max_als = 1e4, bool fast_pI = false, bool direct = true) {
+    double compute_error(ConvClass &converge_test, double tcutCP = 1e-2, ind_t step = 1, ind_t max_rank = 1e5,
```
Comment (evaleev) [inline btas/generic/cp.h:248]: dox

File: btas/generic/converge_class.h [hunk 6]
```diff
@@ -354,5 +358,174 @@ namespace btas {
       return sqrt(abs(nrm));
     }
   };
+
+  template<typename Tensor>
```
Comment (evaleev) [inline btas/generic/converge_class.h:362]: need dox

File: btas/generic/cp.h [hunk 7]
```diff
@@ -23,54 +23,71 @@
...
+
+    // Setting V's where V = A[n]^T A[n]
+    // this means I dont have to recompute these contractions as much.
+    template <typename T, typename Tensor>
+    void set_V(T &t, std::vector<Tensor> &ata) {
+      return;
+    }
+
+    template <typename Tensor>
+    void set_V(FitCheckDF<Tensor> &t, std::vector<Tensor> &ata) {
```
Comment (evaleev) [inline btas/generic/cp.h:86]: dox

File: btas/generic/cp.h [hunk 8]
```diff
@@ -320,8 +334,8 @@ namespace btas{
     /// \returns 2-norm
     /// error between exact and approximate tensor, -1 if calculate_epsilon =
     /// false && ConvClass != FitCheck.
-    virtual double compute_PALS(std::vector <ConvClass> &converge_list, double RankStep = 0.5, size_t panels = 4,
-            int max_als = 20, bool fast_pI = false, bool calculate_epsilon = false,
+    virtual double compute_PALS(std::vector<ConvClass> &converge_list, double RankStep = 0.5, size_t panels = 4,
```
Comment (evaleev) [inline btas/generic/cp.h:337]: dox

File: btas/generic/converge_class.h (hunk 1 above)
Comment (evaleev) [reply btas/generic/converge_class.h:322]: add dox

File: btas/generic/cp_als.h [hunk 9]
```diff
@@ -142,12 +140,12 @@ namespace btas{
     /// error between exact and approximate tensor, -1 if calculate_epsilon =
     /// false && ConvClass != FitCheck.
 
-    double compute_PALS(std::vector <ConvClass> &converge_list, double RankStep = 0.5, size_t panels = 4,
-            int max_als = 20, bool fast_pI = false, bool calculate_epsilon = false,
+    double compute_PALS(std::vector<ConvClass> &converge_list, double RankStep = 0.5, size_t panels = 4,
```
Comment (evaleev) [inline btas/generic/cp_als.h:143]: dox

File: btas/generic/cp_als.h [hunk 10]
```diff
@@ -527,9 +519,8 @@ namespace btas{
     /// \param[in] SVD_initial_guess build inital guess from left singular vectors
     /// \param[in] SVD_rank rank of the initial guess using left singular vector
     /// \param[in] fast_pI Should the pseudo inverse be computed using a fast cholesky decomposition
-    void build_random(ind_t rank, ConvClass &converge_test, bool direct, ind_t max_als,
-                      bool calculate_epsilon, double &epsilon,
-                      bool &fast_pI) override {
+    void build_random(ind_t rank, ConvClass &converge_test, bool direct, ind_t max_als, bool calculate_epsilon,
```
Comment (evaleev) [inline btas/generic/cp_als.h:522]: why pass `fast_pI` by ref?

File: btas/generic/cp_df_als.h [hunk 11]
```diff
@@ -231,11 +233,53 @@ namespace btas{
       return epsilon;
     }
 
-  protected:
-    Tensor &tensor_ref_left;        // Left connected tensor
-    Tensor &tensor_ref_right;       // Right connected tensor
-    size_t ndimL;                      // Number of dimensions in left tensor
-    size_t ndimR;                      // number of dims in the right tensor
+    double compute_cp3_init(ind_t rank, ConvClass converge_test, size_t max_als, bool fast_pI = true,
```
Comment (evaleev) [inline btas/generic/cp_df_als.h:236]: dox

**Review by evaleev — CHANGES_REQUESTED** [review]

File: btas/generic/converge_class.h (hunk 6 above)
Comment (kmp5VT) [reply btas/generic/converge_class.h:362]: This isn't necessary. I thought it was removed, doing that now.

File: btas/generic/cp.h (hunk 7 above)
Comment (kmp5VT) [reply btas/generic/cp.h:86]: Removed

File: btas/generic/cp_als.h (hunk 10 above)
Comment (kmp5VT) [reply btas/generic/cp_als.h:522]: If the user sets fast_pI to true but, during the course of optimization, the Cholesky fails fast_pI is overwritten as false. Keeping a single version and passing by reference allows me to keep track of the variable. An alternative option is to make fast_pI a variable of cp.h

**Review by evaleev — APPROVED** [review]

**Pull request merged**